File: src/core/combat/skills.ts
// 스킬 해소 + 효과 디스패치 (2.5~3.9). 시전자 자기효과 1회 / 대상별 효과 / 동적 재배치.
import type { GameState, Pos, Skill, Unit } from "../types.ts";
import { STATUS_DEFS } from "../../data/statuses.ts";
import { clamp, critPctOf, hasStatus, samePos, unitById } from "../util.ts";
import { computeDamage, dealRawDamage } from "./damage.ts";
import { applyStatusInstance } from "./status.ts";
import { getFormationBonus } from "./formation.ts";
import { areaTargets, computeHitChance } from "./targeting.ts";

/** 앵커(주 대상) 유닛 uid 해소 — targetUid > targetCell/cells[0]의 대상 진영 유닛. 끼어들기 주체 해소에 공유(2.11). */
export function resolveAnchorUid(
  state: GameState,
  actor: Unit,
  skill: Skill,
  sel: { targetUid?: string; targetCell?: Pos; cells?: Pos[] },
): string | undefined {
  if (sel.targetUid) return sel.targetUid;
  if (skill.target === "self") return actor.uid;
  const pos = sel.targetCell ?? sel.cells?.[0];
  if (!pos) return undefined;
  const side = skill.target === "enemy" ? (actor.side === "ally" ? "enemy" : "ally") : actor.side;
  return state.units.find((u) => u.alive && u.side === side && samePos(u.pos, pos))?.uid;
}

function moveUnit(state: GameState, u: Unit, deltaCol: number): void {
  const newCol = clamp(u.pos.col + deltaCol, 0, 3);
  if (newCol === u.pos.col) return;
  const dest: Pos = { row: u.pos.row, col: newCol };
  // 목적지에 살아있는 같은 편 유닛 있으면 이동 취소(막힘)
  const blocked = state.units.some((o) => o.alive && o.side === u.side && o !== u && samePos(o.pos, dest));
  if (blocked) return;
  const from = { ...u.pos };
  u.pos = dest;
  state.log.push({ t: "move", uid: u.uid, from, to: { ...dest } });
}

/** 시전자 자신에게 1회만 적용되는 효과 (광역 중복 방지): applyStatusSelf, move(self). */
function applySelfEffects(state: GameState, actor: Unit, skill: Skill): void {
  for (const eff of skill.effects) {
    if (eff.kind === "applyStatusSelf" && actor.alive) {
      applyStatusInstance(state, actor, actor, eff.statusId, eff.stacks, eff.duration, skill.id);
    } else if (eff.kind === "move" && eff.who === "self" && actor.alive) {
      moveUnit(state, actor, eff.deltaCol);
    }
  }
}

/** 대상별 효과 (광역이면 타겟마다 호출). self 전용 효과는 건너뜀(applySelfEffects가 처리). */
function applyTargetEffects(state: GameState, actor: Unit, skill: Skill, target: Unit, crit: boolean): void {
  for (const eff of skill.effects) {
    switch (eff.kind) {
      case "damage": {
        const atk = getFormationBonus(state, actor, "attackPower");
        const up = actor.skillDmgBonus[skill.id] ?? 0;
        const final = computeDamage(actor, eff.amount + atk + up, crit);
        dealRawDamage(state, target, final, { ignoreShield: hasStatus(actor, "pierce") });
        break;
      }
      case "applyStatus":
        if (target.alive) applyStatusInstance(state, target, actor, eff.statusId, eff.stacks, eff.duration, skill.id);
        break;
      case "cleanse": {
        const before = target.statuses.length;
        target.statuses = target.statuses.filter((s) => STATUS_DEFS[s.defId].buff);
        if (target.statuses.length !== before) state.log.push({ t: "cleanse", targetUid: target.uid });
        break;
      }
      case "shield": {
        const def = getFormationBonus(state, actor, "defensePower");
        const amt = Math.round(eff.amount + def) + target.equipShieldGainAdd; // 방어구 쉴드 획득량 보정(받는 쪽, 4.3)
        target.shield += amt;
        state.log.push({ t: "shieldGain", targetUid: target.uid, amount: amt });
        break;
      }
      case "heal": {
        const def = getFormationBonus(state, actor, "defensePower");
        const before = target.hp;
        target.hp = Math.min(target.hpMax, target.hp + Math.round(eff.amount + def));
        state.log.push({ t: "heal", targetUid: target.uid, amount: target.hp - before });
        break;
      }
      case "move":
        if (eff.who === "target" && target.alive) moveUnit(state, target, eff.deltaCol);
        break;
      // applyStatusSelf, move(self) → applySelfEffects에서 1회 처리
    }
  }
}

export function resolveSkill(
  state: GameState,
  actor: Unit,
  skill: Skill,
  sel: { targetUid?: string; targetCell?: Pos; cells?: Pos[] },
): void {
  state.log.push({ t: "skillUsed", uid: actor.uid, skillId: skill.id, targetUid: sel.targetUid });

  // 시전자 자기 효과 1회 (광역 중복 방지)
  applySelfEffects(state, actor, skill);

  // 앵커 칸: 명시 칸 > 대상 유닛 위치 > 시전자 위치
  const anchor: Pos = sel.targetCell ?? (sel.targetUid ? unitById(state, sel.targetUid).pos : actor.pos);
  const targets = areaTargets(state, actor, skill, anchor, sel.cells);

  for (const tgt of targets) {
    if (!tgt.alive) continue;
    if (skill.target === "enemy") {
      const chance = computeHitChance(actor, skill, tgt);
      if (!skill.alwaysHit && !state.rng.chance(chance)) {
        state.log.push({ t: "miss", uid: actor.uid, targetUid: tgt.uid, chance });
        continue;
      }
      const crit = state.rng.chance(critPctOf(actor));
      state.log.push({ t: "hit", uid: actor.uid, targetUid: tgt.uid, chance, crit });
      applyTargetEffects(state, actor, skill, tgt, crit);
    } else {
      applyTargetEffects(state, actor, skill, tgt, false); // self/ally: 명중 판정 없음
    }
  }
}

File: rust/spr-core/src/skills.rs
//! 스킬 해소 + 효과 디스패치 (TS `combat/skills.ts`, 2.5~3.9). 자기효과 1회 / 대상별 효과 / 동적 재배치.
//! 패시브 훅(skillUsed/onHit/onMiss/onHeal/onShieldGain/onMove 등)은 9f 슬라이스서 — 현재 효과 무발생(데모/테스트는 패시브-프리).
use crate::damage::{compute_damage, deal_raw_damage};
use crate::formation::get_formation_bonus;
use crate::status::apply_status_instance;
use crate::targeting::{area_targets, compute_hit_chance};
use crate::util::{crit_pct_of, has_status, StatusDefs};
use spr_types::combat::{GameEvent, GameState};
use spr_types::data::Pos;
use spr_types::skills::{Skill, SkillEffect};

fn idx_of_uid(state: &GameState, uid: &str) -> usize {
    state.units.iter().position(|u| u.uid == uid).expect("unit not found")
}

/// 동적 재배치(6.4): deltaCol만큼 이동(0~3 클램프), 같은 편 점유 칸이면 취소.
fn move_unit(state: &mut GameState, unit_idx: usize, delta_col: i64) {
    let (new_col, from, side, uid) = {
        let u = &state.units[unit_idx];
        ((u.pos.col + delta_col).clamp(0, 3), u.pos, u.side.clone(), u.uid.clone())
    };
    if new_col == from.col {
        return;
    }
    let dest = Pos { row: from.row, col: new_col };
    let blocked = state.units.iter().enumerate().any(|(i, o)| i != unit_idx && o.alive && o.side == side && o.pos == dest);
    if blocked {
        return;
    }
    state.units[unit_idx].pos = dest;
    state.log.push(GameEvent::Move { uid, from, to: dest });
    // fireTrigger(onMove, enterCell) — 9f서.
}

/// 시전자 자기효과 1회: applyStatusSelf, move(self).
fn apply_self_effects(state: &mut GameState, actor_idx: usize, skill: &Skill) {
    for eff in &skill.effects {
        match eff {
            SkillEffect::ApplyStatusSelf { status_id, stacks, duration } if state.units[actor_idx].alive => {
                let uid = state.units[actor_idx].uid.clone();
                apply_status_instance(state, actor_idx, &uid, status_id, *stacks, *duration, Some(skill.id.clone()));
            }
            SkillEffect::Move { who, delta_col } if who == "self" && state.units[actor_idx].alive => {
                move_unit(state, actor_idx, *delta_col);
            }
            _ => {}
        }
    }
}

/// 대상별 효과(광역=타겟마다). self 전용 효과는 건너뜀(apply_self_effects 처리).
fn apply_target_effects(state: &mut GameState, actor_idx: usize, skill: &Skill, target_idx: usize, crit: bool, defs: &StatusDefs) {
    for eff in &skill.effects {
        match eff {
            SkillEffect::Damage { amount } => {
                let (final_dmg, ignore_shield) = {
                    let actor = &state.units[actor_idx];
                    let atk = get_formation_bonus(state, actor, "attackPower");
                    let up = *actor.skill_dmg_bonus.get(&skill.id).unwrap_or(&0);
                    (compute_damage(actor, amount + atk + up, crit, defs), has_status(actor, "pierce"))
                };
                deal_raw_damage(state, target_idx, final_dmg, ignore_shield, defs);
            }
            SkillEffect::ApplyStatus { status_id, stacks, duration } => {
                if state.units[target_idx].alive {
                    let auid = state.units[actor_idx].uid.clone();
                    apply_status_instance(state, target_idx, &auid, status_id, *stacks, *duration, Some(skill.id.clone()));
                }
            }
            SkillEffect::Cleanse => {
                let t = &mut state.units[target_idx];
                let before = t.statuses.len();
                t.statuses.retain(|s| defs.get(&s.def_id).map(|d| d.buff).unwrap_or(false));
                if t.statuses.len() != before {
                    let uid = t.uid.clone();
                    state.log.push(GameEvent::Cleanse { target_uid: uid });
                }
            }
            SkillEffect::Shield { amount } => {
                let amt = {
                    let actor = &state.units[actor_idx];
                    let def = get_formation_bonus(state, actor, "defensePower");
                    amount + def + state.units[target_idx].equip_shield_gain_add
                };
                let t = &mut state.units[target_idx];
                t.shield += amt;
                let uid = t.uid.clone();
                state.log.push(GameEvent::ShieldGain { target_uid: uid, amount: amt });
                // fireTrigger(onShieldGain) — 9f서.
            }
            SkillEffect::Heal { amount } => {
                let add = {
                    let actor = &state.units[actor_idx];
                    get_formation_bonus(state, actor, "defensePower")
                };
                let t = &mut state.units[target_idx];
                let before = t.hp;
                t.hp = (t.hp + amount + add).min(t.hp_max);
                let gained = t.hp - before;
                let uid = t.uid.clone();
                state.log.push(GameEvent::Heal { target_uid: uid, amount: gained });
                // fireTrigger(onHeal) — 9f서.
            }
            SkillEffect::Move { who, delta_col } => {
                if who == "target" && state.units[target_idx].alive {
                    move_unit(state, target_idx, *delta_col);
                }
            }
            SkillEffect::ApplyStatusSelf { .. } => {} // apply_self_effects가 처리
        }
    }
}

/// 스킬 해소 진입점. TS resolveSkill. sel = (target_uid, target_cell, cells).
pub fn resolve_skill(
    state: &mut GameState,
    actor_idx: usize,
    skill: &Skill,
    target_uid: Option<&str>,
    target_cell: Option<Pos>,
    cells: Option<&[Pos]>,
    defs: &StatusDefs,
) {
    let actor_uid = state.units[actor_idx].uid.clone();
    state.log.push(GameEvent::SkillUsed {
        uid: actor_uid.clone(),
        skill_id: skill.id.clone(),
        target_uid: target_uid.map(|s| s.to_string()),
    });
    // fireTrigger(skillUsed) — 9f서.

    apply_self_effects(state, actor_idx, skill);

    // 앵커 칸: 명시 칸 > 대상 유닛 위치 > 시전자 위치
    let anchor = if let Some(tc) = target_cell {
        tc
    } else if let Some(tu) = target_uid {
        state.units[idx_of_uid(state, tu)].pos
    } else {
        state.units[actor_idx].pos
    };
    let free_cells: Vec<Pos> = cells.map(|c| c.to_vec()).unwrap_or_default();
    let targets = area_targets(state, actor_idx, skill, anchor, &free_cells);

    for target_idx in targets {
        if !state.units[target_idx].alive {
            continue;
        }
        if skill.target == "enemy" {
            let chance = compute_hit_chance(&state.units[actor_idx], skill, &state.units[target_idx]);
            let tuid = state.units[target_idx].uid.clone();
            if !skill.always_hit && !state.rng.chance(chance) {
                state.log.push(GameEvent::Miss { uid: actor_uid.clone(), target_uid: tuid, chance });
                // fireTrigger(onMiss) — 9f서.
                continue;
            }
            let crit = state.rng.chance(crit_pct_of(&state.units[actor_idx], defs));
            state.log.push(GameEvent::Hit { uid: actor_uid.clone(), target_uid: tuid, chance, crit });
            // fireTrigger(onHit) — 9f서.
            apply_target_effects(state, actor_idx, skill, target_idx, crit, defs);
        } else {
            apply_target_effects(state, actor_idx, skill, target_idx, false, defs);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::battle::create_battle;
    use spr_types::canonical::canonical_json;
    use spr_types::data::{Encounter, Placement};

    // 패시브-프리 시나리오: thug(무특성/무패시브) vs thug → fire_trigger no-op이 TS와 바이트 동일.
    // 효과 다양성(applyStatus/shield/heal/cleanse/move)·패시브 발화는 9h 풀 differential서 검증.
    fn thug_duel() -> Encounter {
        Encounter {
            id: "t".into(),
            name: "t".into(),
            allies: vec![Placement { char_id: "thug".into(), pos: Pos { row: 0, col: 0 } }],
            enemies: vec![Placement { char_id: "thug".into(), pos: Pos { row: 0, col: 0 } }],
            boss: false,
        }
    }

    fn run(seed: u32) -> String {
        let chars = spr_data::characters();
        let skills = spr_data::skills();
        let defs = spr_data::status_defs();
        let mut s = create_battle(seed, &thug_duel(), &chars);
        let a = s.units.iter().position(|u| u.uid == "a0_thug").unwrap();
        let base = s.log.len();
        resolve_skill(&mut s, a, &skills["thug_punch"], Some("e0_thug"), None, None, &defs);
        canonical_json(&s.log[base..])
    }

    #[test]
    fn resolve_skill_hit_parity() {
        // seed1: 명중 비크리. base 6 + 포메이션 atk4(단독 col0) = 10. RNG 2소비(명중·크리).
        assert_eq!(
            run(1),
            r#"[{"skillId":"thug_punch","t":"skillUsed","targetUid":"e0_thug","uid":"a0_thug"},{"chance":79,"crit":false,"t":"hit","targetUid":"e0_thug","uid":"a0_thug"},{"base":10,"final":10,"t":"damage","targetUid":"e0_thug","toHp":10,"toShield":0}]"#
        );
    }

    #[test]
    fn resolve_skill_miss_parity() {
        // seed7: 빗나감 → miss 이벤트만(RNG 1소비). hit 분기 RNG 순서 일치 확인.
        assert_eq!(
            run(7),
            r#"[{"skillId":"thug_punch","t":"skillUsed","targetUid":"e0_thug","uid":"a0_thug"},{"chance":79,"t":"miss","targetUid":"e0_thug","uid":"a0_thug"}]"#
        );
    }
}
